### ISSUE TITLE
GCP上での環境構築

### ISSUE THREAD

**Issue by hariwasabi** [sysh333/soundnotes#11]
@sysh333 

お疲れ様です。
辻本です。

GCP上で環境構築していますが、設定内容不備があるようでちょっと手こずっております。
すみませんが、もう少しお時間ください。

よろしくお願いいたします。

**Comment by hariwasabi** [comment]
@sysh333 

GCPでGCEとcloud sqlを使って画面表示できるようになりました。
ただ、エラーが出てしまうので、確認したところ、ローカル環境でも同様のエラーがでていました。

最初に詳細にみておくべきでした。すみません。

ちょっとエラーを解決しないと先に進めなさそうですので調べます。
もし何か思いつくことありましたらご教授いただければ幸いです。

**Comment by sysh333** [comment]
辻本さま

孫です。お世話になります。
ローカル環境でのエラーはどんなものでしょうか。

こちらでは出ていない感じがしますが、
SQL関係のエラーでしょうか？

**Comment by hariwasabi** [comment]
ぱっと見た限りはSQLのエラーだと思います。

あとで、エラー内容をお伝えします！

**Comment by sysh333** [comment]
こうゆうやつですか？

[1] Error: Incorrect datetime value: '2018-07-09T08:17:10.736Z' for column 'submit_time' at row 1
[1]     at PromiseConnection.query (/Users/yson/dev/myapp/SoundNotes/server/node_modules/mysql2/promise.js:89:20)
[1]     at router.post (/Users/yson/dev/myapp/SoundNotes/server/note.route.js:31:39)
[1]     at <anonymous>
[1]     at process._tickCallback (internal/process/next_tick.js:188:7)

**Comment by sysh333** [comment]
Sqlの５.６以降でsql_modeにSTRICT_TRANS_TABLESが指定されるとのことで起きる事象

**Comment by sysh333** [comment]
http://onlineconsultant.jp/pukiwiki/?MySQL5.6%E3%81%AB%E3%81%99%E3%82%8B%E3%81%A8%E3%82%A8%E3%83%A9%E3%83%BC1292%20-%20Incorrect%20datetime%20value%E3%81%8C%E7%99%BA%E7%94%9F%E3%81%99%E3%82%8B

**Comment by hariwasabi** [comment]
そうです。
これは出ています。

他にもあったとおもいます。

**Comment by sysh333** [comment]
上記の直し方は２つあり、私は簡易な方でやりました。

/usr/local/etcの中にあるmy.cnfの中身
# Default Homebrew MySQL server config
[mysqld]
# Only allow connections from localhost
bind-address = 127.0.0.1

ここに、sql_mode=‘’　を追記して、SQLを再起動すると　少なくともローカルでも問題なく動きます。

**Comment by sysh333** [comment]
上記はGCP上で、どうにか直せるか、わかっていません。

**Comment by sysh333** [comment]
それでもダメだったら、、、ソースコードをいじる必要が出てきそうですね。

**Comment by hariwasabi** [comment]
mysql client からsql modeを変更できるんですね。
なるほどです。

**Comment by hariwasabi** [comment]
試してみます。

**Comment by sysh333** [comment]
GCP上で出ているエラーも
Error: Incorrect datetime value:
でしょうか？

**Comment by sysh333** [comment]
これを頑張ってローカルで動作せせても、GCP上で全然違うエラーが出てそうで怖い。

**Comment by hariwasabi** [comment]
GCPでもこのエラーはでています。
エラー出たので、ローカルでもやってみたら、あれ？動いてない。。と気づいた流れです。

Mysql client で設定できるならそれがよいかと。

まずはここをクリアします

**Comment by hariwasabi** [comment]
@sysh333 
ローカルは動きました。
ただ、同様の設定がGCPでなかなかできずに考え中です。
・my.conf　→　GCPでこのファイル見つからず
・mysql clientでの設定　→　global.sqlmodeを変更したいが、SUPER権限がない

取り急ぎのご報告です。

**Comment by sysh333** [comment]
ありがとうございます。
権限の問題、どうにかなりそうでしょうか？

**Comment by hariwasabi** [comment]
@sysh333 

権限の問題、今の所は解決できていないです。

**Comment by hariwasabi** [comment]
@sysh333 
GCP公式のcloud sql(mysql)のドキュメントをみる限りだと、rootユーザでもなぜかSuper権限が付与されないらしく、厳しいかもしれませんね。

node.js側で、mysqlのDATATIMEのフォーマットに変換するとか可能でしょうか？

**Comment by sysh333** [comment]
はい、おそらくこれはデプロイ側（GCP上）でどうにかする問題ではなく、ソースコードの方のフォーマット変換とか頑張る必要があると感じました。（あってます？）

デプロイは一旦stopで、近いうちにソースコード直したら、もう一度連絡してもいいですか？

**Comment by hariwasabi** [comment]
@sysh333 

はい、もちろん大丈夫です！お願いいたします。
local環境構築時もmysqlの設定変更が必要となっているので、ソース修正することが一番いいかと思いました。

遅れてしまい、申し訳ございませんがよろしくお願いいたします。

**Comment by sysh333** [comment]
@hariwasabi 
孫です。モードの変更はこれを見ればできるようです。
https://cloud.google.com/sql/docs/mysql-flags?hl=ja

実際に設定した人のブログ↓
https://introvertedengineer.com/2017/03/23/mysql-sql_mode-troubles-google-cloud-sql/
ソースは、今大幅に書き換えをしようとしているので、

結論として、TRADITIONAL または、ALLOW_INVALID_DATESのflagをつければ大丈夫のようです。

本件、かなり延び延び担っているので、早く終わらせるという意味でこれを設定いただき、
全体のデプロイを一度実演いただければ、本件クローズとさせて欲しいです。
（手順とか書くのはめんどくさいですよね。　一回一緒に、私の環境でデプロイすればOKです）

**Comment by hariwasabi** [comment]
@sysh333 

孫さん

ご連絡ありがとうございます。
モード変更の情報、一度試してみます。

本日、時間がとれないので、明日にやってご報告いたします！

よろしくお願いいたします。

**Comment by sysh333** [comment]
@hariwasabi 
承知しました。
ちなみに、それでデプロイできなくても、今までやっていただいた手順を教えてもらうということで、
本件クローズで大丈夫です。
明日以降、お時間が合えば、教えていただければと思います。（何森さんにもメールしました）

**Comment by hariwasabi** [comment]
@sysh333 

承知しました！
何森さんが見れないので、メールのほうでご報告いたしますね。
